### DISCUSSION TITLE
Make Dagger GHA pipelines more Dagger friendly

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12432] [maintainers]
context from a random reddit thread: 

![image](https://user-images.githubusercontent.com/1578458/217579621-fa316f45-7ee0-4cc3-bc08-e7d7635d9299.png)

quoting @shykes: 

> Thinking about it more, I guess converting everything to dagger except one workflow might be the ideal balance. I know @aluzzardi had thoughts on this. Maybe we make this an issue and get everyone’s thoughts there? 

Regardless of the reddit drama, it's somewhat true that some users will probably take a look at our `dagger/dagger` GHA pipelines and use them as a reference to setup their projects. The way that they're architected still has a lot of coupling to GHA in particular our  main "publish" (https://github.com/dagger/dagger/blob/main/.github/workflows/publish.yml) one.

This issue serves as a starting point to understand if/what changes it would be possible to change in our `dagger/dagger` GHA pipelines to bring them closer to a Dagger native way.

**Comment by shykes** [comment]
I think an actionable first step would be to document our own CI setup with comments and a README, to provide more context. Not just to explain what, but also why. 

Especially for the "dagger sandwich" workflows like publish, where we still have lots of yaml spaghetti, but for pragmatic reasons that will resonate with experienced teams. One of them being: "this works well enough with an off-the-shelf github action, and is not worth the effort of porting yet. But that's ok since we can mix dagger and non-dagger pipelines seamlessly".

I think we should do this before any other more ambitious changes.

**Comment by marcosnils** [comment]
:+1: I can take the ownership here and send a PR with a proposal.
